feature(Tax Ipr Confuguration)

lomamech opened the pull request:
- Implemente taxe IPR management and IPR Configuration Management
- Implement component bhIprScale which allows to select the scale IPR to configure
- Implement Tax IPR Configuration Integration Test
- Implement test E2E for tax IPR management
- Implement test E2E of tax Ipr configuration of tranches

reference #2292

lomamech committed:
perf(Tax IPR Management)
- Modification of the tax ipr table and addition of the tax ipr
  configuration table in order to have the possibility of defining the
  configuration of the IPR bands to be used during the calculation of the
  IPR tax

lomamech committed:
test(Tax Ipr Configuration)
- Implement Tax IPR Configuration Integration Test

lomamech committed:
test(Tax Ipr Configuration)
- Implement test E2E for tax IPR management

lomamech committed:
test(Tax IPR Configuration)
- Implement test E2E of tax Ipr configuration of tranches`

jniles commented on client/src/js/components/bhIprScale.js:
```diff
@@ -0,0 +1,45 @@
+    ctrl.onUpdate({ scaleId : scale.id });
+  };
+
+  function loadDefaultCurrency() {
```
Should this be called `loadDefaultCurrency()`?  It looks like it doesn't have anything to do with currencies...

jniles commented on client/src/modules/ipr_tax/configuration/iprTaxConfig.html:
```diff
@@ -0,0 +1,51 @@
+
+    <div class="toolbar">
+      <div class="toolbar-item">
+        <button ng-disabled="!IprTaxConfigCtrl.taxIprId" class="btn btn-default text-capitalize" ui-sref="configuration.createConfig({ taxIprId : IprTaxConfigCtrl.taxIprId })" data-method="create">
```
Nice `uiSref` call! 👍

jniles commented on client/src/modules/ipr_tax/configuration/iprTaxConfig.js:
```diff
@@ -0,0 +1,127 @@
+
+  // update an existing IprTaxConfig
+  function editIprTaxConfig(ipr) {
+    $state.go('configuration.editConfig', { taxIprId : vm.taxIprId, id : ipr.id });
```
👍

jniles commented on client/src/modules/ipr_tax/ipr_tax.html:
```diff
@@ -0,0 +1,44 @@
+      </div>
+
+      <div class="toolbar-item">
+        <button
```
Just a note: we use this so often, we might as well make a component for it.  It shouldn't be included in this PR, but we can definitely plan it as future work.

jniles commented on client/src/modules/ipr_tax/ipr_tax.routes.js:
```diff
@@ -0,0 +1,73 @@
+        templateUrl : 'modules/ipr_tax/ipr_tax.html',
+      })
+
+      .state('configuration', {
```
The `configuration` state is a really generic name, and we may run into problems if another developer makes the same state for something like inventory configuration.   Could you name it `taxConfiguration`, `iprConfiguration` something similar?  That way we know what it is for?

jniles commented on client/src/modules/ipr_tax/ipr_tax.routes.js:
```diff
@@ -0,0 +1,73 @@
+      })
+
+      .state('configuration.editConfig', {
+        url : '/:taxIprId/:id/configuration/edit',
```
This url is probably better arranged like this:
```js
 url : '/:taxIprId/configuration/:id/edit'
```
that way, you know that the `:id` belongs to the `configuration`.

jniles commented on client/src/modules/ipr_tax/ipr_tax.service.js:
```diff
@@ -0,0 +1,17 @@
+ * @description
+ * Encapsulates common requests to the /iprTaxes/ URL.
+ */
+function IprTaxService(Api, Modal) {
```
It doesn't look like you need `$uibModal`, you can remove it.

jniles commented on client/src/modules/ipr_tax/ipr_tax_config.service.js:
```diff
@@ -0,0 +1,47 @@
+ * @description
+ * Encapsulates common requests to the /iprTaxConfig/ URL.
+ */
+function IprTaxConfigService(Api, Modal) {
```
There is no need for the Modal $inject here.

jniles commented on client/src/js/components/bhIprScale.js:
```diff
@@ -0,0 +1,45 @@
+        ctrl.iprTaxes = new Store();
+        ctrl.iprTaxes.setData(iprTaxes);
+        loadDefaultCurrency();
+      });
```
I think this should have error handling.

jniles commented on client/src/modules/ipr_tax/modals/ipr_tax_config.modal.html:
```diff
@@ -0,0 +1,70 @@
+    <div class="modal-body" ng-class="{ 'has-error' : IprTaxForm.$submitted && IprTaxForm.tranche_annuelle_debut.$invalid }">
+      <label class="control-label" translate>FORM.LABELS.ANNUAL_TRANCH_FROM</label>
+      <div class="input-group">
+        <span class="input-group-addon text-uppercase"> {{ IprTaxConfigModalCtrl.iprTax.symbol }} </span>
```
Can this use the `bhCurrencyInput`?  It might be worth making the `bhCurrencyInput` flexible enough to work for this case - it would save a lot of space!

jniles commented on client/src/modules/ipr_tax/templates/impot_annuel.tmpl.html:
```diff
@@ -0,0 +1,3 @@
+<div class="ui-grid-cell-contents text-right">
+  {{ row.entity.impot_annuel | currency: row.entity.currency_id }}
```
All of these cell templates will take a long time to render.  It's best to define this using the column definitions for better performance.  This would be:
```js
{
  field: 'impot_annuel',
  cellClass : 'text-right',
  cellFilter : 'currency:row.entity.currency_id',
}
```

Try this for all the rest of the cells as well.

jniles commented on client/src/modules/ipr_tax/templates/rate.tmpl.html:
```diff
@@ -0,0 +1,3 @@
+<div class="ui-grid-cell-contents text-right">
+  {{ row.entity.rate }} %
```
You could actually make a filter for this.  [See this for example](https://gist.github.com/jeffjohnson9046/9470800).

jniles commented on client/src/modules/templates/bhIprScale.tmpl.html:
```diff
@@ -0,0 +1,13 @@
+<div class="btn-group" uib-dropdown dropdown-append-to-body>
```
Nice 👍

jniles commented on client/src/modules/templates/bhIprScale.tmpl.html:
```diff
@@ -0,0 +1,13 @@
+<div class="btn-group" uib-dropdown dropdown-append-to-body>
+  <a class="btn btn-default" uib-dropdown-toggle>
```
If this is a dropdown, can you append the `bh-dropdown-auto-dropup` tag to this?  It will make sure it behaves correctly when at the bottom of the list.

jniles reviewed (commented):
Just a few comments here.  I'll be able to give a more full review tomorrow!

lomamech committed:
refactor(Tax Ipr Configuration)
- Fix and rename function name
- Rename route, use iprConfiguration instead configuration
- user better arrangement for the URL
- Use bhCurrencyImput dor the setting for Annual Tranch of IPR scale
- Remove all not necessary template and use currency filter
- add bh-dropdown-auto-dropup
- Update and Improve bhCurrencyInput
- Update test E2E for iprTax Configuration

lomamech committed:
Fix Conflict with Branch Master

jniles reviewed (approved):
I worry this will get out of date, so we'll go ahead an merge this.  It looks pretty good to me!

bors r+
